src/psyclone/configuration.py

```diff
@@ -0,0 +1,426 @@
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: R. W. Ford and A. R. Porter, STFC Daresbury Lab
+
+'''
+PSyclone configuration management module.
+
+Deals with reading the config file and storing default settings.
+'''
+
+from __future__ import absolute_import
+import os
+
+
+# Name of the config file we search for
+_FILE_NAME = "psyclone.cfg"
+
+
+class ConfigurationError(Exception):
+    '''
+    Class for all configuration-related errors.
+    '''
+    def __init__(self, value, config=None):
+        '''
+        :param str value: error message
+        :param config: the Config object associated with the error
+        :type :py:class:`psyclone.configuration.Config`
+        '''
+        Exception.__init__(self, value)
+        self.value = "PSyclone configuration error"
+        if config:
+            self.value += " (file={0})".format(config.filename)
+        self.value += ": "+value
+
+    def __str__(self):
+        return repr(self.value)
+
+
+def _str_to_list(svalue):
+    '''
+    Helper routine to take a string containing a list of values and return
+    a list of those values.
+
+    :param str svalue: string containing space- or comma-delimited list
+                       of items
+    :returns: list of strings
+    :rtype: list
+    '''
+    if "," in svalue:
+        # Comma delimited
+        return [
+            str(item.strip()) for item in svalue.split(",")
+            if item.strip() != '']
+    # Space delimited
+    return [
+        str(item.strip()) for item in svalue.split(" ")
+        if item.strip() != '']
+
+
+class ConfigFactory(object):
+    '''
+    Create our singleton Config object. If config_file is specified
+    then we throw-away the old Config and create a new one.
+
+    :param str config_file: Specific configuration file to use when \
+                            creating the Config object.
+    '''
+    _instance = None  # Our single Config object
+
+    def __init__(self, config_file=None):
+        if not ConfigFactory._instance or config_file:
```

Comment:
For consistency you should have the arguments described here not in the class description